[Feature Request]: Support Type Inference on Python 3.12

### What would you like to happen?

Part of #29149

Beam type inference tests currently fail under Python 3.12:

Repro:

1. Clone: https://github.com/apache/beam/pull/30828
2. Install Beam + test dependencies
3. Run: pytest apache_beam/typehints/

Several tests fail:

```
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testAlwaysReturnsEarly - AssertionError: <class 'int'> != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testBuildTupleUnpack - AssertionError: Tuple[Union[<class 'int'>, <class 'str'>], ...] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testBuildTupleUnpackWithCall - AssertionError: Tuple[Union[<class 'float'>, <class 'int'[67 chars]r'>]] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testCallFunctionOnAny - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testCallNestedLambda - AssertionError: Tuple[63 chars]ferenceTest.testCallNestedLambda.<locals>.Foo'>, <class 'int'>] != Tuple[63 chars]ferenceTest.testCallNe...
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testCallNullaryMethod - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testDepthMethod - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testDictComprehension - AssertionError: Dict[Any, Any] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testDictComprehensionSimple - AssertionError: Dict[<class 'str'>, <class 'int'>] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testGenerator - AssertionError: Iterable[<class 'int'>] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testGeneratorComprehension - AssertionError
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testGetAttr - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testGetItemSlice - AssertionError: List[<class 'str'>] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testJumpOffsets - AssertionError: List[<class 'int'>] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testListComprehension - AssertionError: List[<class 'int'>] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testMethod - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testNoneReturn - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testRow - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testSet - AssertionError: Set[Union[]] != Any
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testSetAttr - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testSetDeleteGlobal - IndexError: tuple index out of range
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testTupleListComprehension - AssertionError: List[<class 'int'>] != Any
```


### Issue Priority

Priority: 2 (default / most feature requests should be filed as P2)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

A few tests are still failing: https://github.com/apache/beam/actions/runs/9010406521/job/24756550768?pr=31223

```
FAILED apache_beam/typehints/intrinsic_one_ops_test.py::IntrinsicOneOpsTest::test_unary_intrinsic_ops_are_in_the_same_order_as_in_cpython
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testDictComprehension
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testDictComprehensionSimple
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testListComprehension
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testSet
FAILED apache_beam/typehints/trivial_inference_test.py::TrivialInferenceTest::testTupleListComprehension
FAILED apache_beam/transforms/batch_dofn_test.py::BatchDoFnParameterizedTest_NoReturnAnnotation::test_get_output_batch_type
=== 33 failed, 5776 passed, 1563 skipped, 2896 warnings in 719.14s (0:11:59) ===
```

--- a/sdks/python/apache_beam/typehints/intrinsic_one_ops_test.py
+++ b/sdks/python/apache_beam/typehints/intrinsic_one_ops_test.py
@@ -30,9 +30,7 @@ class IntrinsicOneOpsTest(unittest.TestCase):
   def test_unary_intrinsic_ops_are_in_the_same_order_as_in_cpython(self):
     if sys.version_info >= (3, 12):
       dis_order = dis.__dict__['_intrinsic_1_descs']
-      beam_ops = [fn.__name_upper() for fn in intrinsic_one_ops.INT_ONE_OPS]
-      for fn in intrinsic_one_ops.INT_ONE_OPS:
-        beam_ops.append(fn.__name__.upper())
+      beam_ops = [fn.__name__.upper() for fn in intrinsic_one_ops.INT_ONE_OPS]
       self.assertListEqual(dis_order, beam_ops)
 
 

--- a/sdks/python/apache_beam/typehints/opcodes.py
+++ b/sdks/python/apache_beam/typehints/opcodes.py
@@ -471,7 +471,7 @@ load_fast_check = load_fast
 
 def load_fast_and_clear(state, arg):
   state.stack.append(state.vars[arg])
-  del state.vars[arg]
+  state.vars[arg] = None
 
 
 def store_fast(state, arg):

